server/lib/parties/party-service.ts

```diff
@@ -332,7 +332,13 @@ export default class PartyService {
       })
     }
 
-    this.clientSocketsToPartyId.delete(clientSockets)
+    const clientPartyId = this.clientSocketsToPartyId.get(clientSockets)!
```

Comment:
If this wasn't caught by tests, then test coverage is not sufficient. Thus, this fix should include an additional test or a modification to existing tests

Reply:
I wasn't testing this because `clientSocketsToPartyId` is a private property in the party service. But I guess it doesn't make sense to have private stuff in services when writing unit tests, since we most likely want to test the internals of the service?

Reply:
Generally, no. The service should be testable by controlling inputs/examining outputs. Stuff that is "private to the service" is an implementation detail, and shouldn't really be depended upon for tests.

(This isn't always possible, but in new code that has tests written for it directly, it is nearly always possible to design things in that way)

Think more deeply about what the issue you're seeing here is and how you would detect that in a test. What test could you write that would fail with the old code but succeed with the new code? (Then ideally, delete this fix, run the test, ensure it fails, and then re-add the fix and ensure it passes)

Reply:
I'll just give you an example so that we don't go back and forth on this forever and actually get this done:

The issue you're seeing here is that in the case of switching parties, a user ends up only halfway into a party, because after they switched parties the system comes along and deletes their registration? How would this manifest in the application, what error states would occur? What did you see when this happened that even made you realize it was an issue?

I'd guess the user would be unable to leave the new party because the system would fail to look up their client ID. Thus a reasonable test might be:

1) Join a party
2) Switch to a different party
3) Try to leave the different party
4) Verify that leaving the party worked

Think less about this stuff being "I want to call function X, Y, Z" and more about actions that users are taking, and how those actions relate to the APIs present. The more realistic the sequence to what users are doing, and the more closely we verify state that is actually being used by those users, the better our tests will be at finding "real" issues and the more resilient they'll be to changes in the underlying implementation.

Reply:
Actually, the user was able to leave the new party just fine; the problem occurred when trying to get the party by client's sockets (after it was erroneously deleted), which was mostly used in cleaning up the user when they disconnected (I only even noticed this bug while getting the error logged in the console). But anyway, added the test for that too.

For what is worth, should probably add tests at various other places to make sure stuff gets cleaned up properly when a user arbitrarily disconnects. But gonna do that outside this PR. The reason why I didn't do it in the first place is because I wasn't sure how to simulate the user disconnect with our fake nydus client we use in testing, but after spending two hours on it now, I finally figured it out! And it's one line of code ><

Reply:
Some lines of code are more expensive than others :P